services/pushover/service.go

```diff
@@ -0,0 +1,290 @@
+)
+
+type Service struct {
+	configValue atomic.Value
+	logger      *log.Logger
+}
+
+func NewService(c Config, l *log.Logger) *Service {
+	s := &Service{
+		logger: l,
+	}
+	s.configValue.Store(c)
+	return s
+}
+
+func (s *Service) Open() error {
+	return nil
+}
+
+func (s *Service) Close() error {
+	return nil
+}
+
+func (s *Service) Update(newConfig []interface{}) error {
+	if l := len(newConfig); l != 1 {
+		return fmt.Errorf("expected only one new config object, got %d", l)
+	}
+	if c, ok := newConfig[0].(Config); !ok {
+		return fmt.Errorf("expected config object to be of type %T, got %T", c, newConfig[0])
+	} else {
+		s.configValue.Store(c)
+	}
+	return nil
+}
+
+type testOptions struct {
+	User      string      `json:"user"`
+	Message   string      `json:"message"`
+	Device    string      `json:"device"`
+	Title     string      `json:"title"`
+	URL       string      `json:"url"`
+	URLTitle  string      `json:"url_title"`
+	Sound     string      `json:"sound"`
+	Timestamp bool        `json:"timestamp"`
+	Level     alert.Level `json:"level"`
+}
+
+func (s *Service) TestOptions() interface{} {
+	c := s.config()
+	return &testOptions{
+		User:    c.User,
+		Message: "test pushover message",
+		Level:   alert.Critical,
+	}
+}
+
+func (s *Service) Test(options interface{}) error {
+	o, ok := options.(*testOptions)
+	if !ok {
+		return fmt.Errorf("unexpected options type %t", options)
+	}
+
+	return s.Alert(
+		o.User,
+		o.Message,
+		o.Device,
+		o.Title,
+		o.URL,
+		o.URLTitle,
+		o.Sound,
+		o.Timestamp,
+		o.Level,
+	)
+}
+
+func (s *Service) config() Config {
+	return s.configValue.Load().(Config)
+}
+
+func (s *Service) Alert(user, message, device, title, URL, URLTitle, sound string, timestamp bool, level alert.Level) error {
+	url, post, err := s.preparePost(user, message, device, title, URL, URLTitle, sound, timestamp, level)
+	if err != nil {
+		return err
+	}
+
+	resp, err := http.PostForm(url, post)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		type response struct {
+			Error string `json:"error"`
+		}
+		r := &response{Error: fmt.Sprintf("failed to understand Slack response. code: %d content: %s", resp.StatusCode, string(body))}
+		b := bytes.NewReader(body)
+		dec := json.NewDecoder(b)
+		dec.Decode(r)
+		return errors.New(r.Error)
+	}
+
+	return nil
+}
+
+// priority returns the pushover priority as defined by the Pushover API
+// documentation https://pushover.net/api
+func priority(level alert.Level) int {
+	switch level {
+	case alert.OK:
+		// send as -2 to generate no notification/alert
+		return -2
+	case alert.Info:
+		// -1 to always send as a quiet notification
+		return -1
+	case alert.Warning:
+		// 1 to display as high-priority and bypass the user's quiet hours,
+		return 1
+	case alert.Critical:
+		// 2 to also require confirmation from the user
+		return 2
+	}
```

Comment:
This mapping should be documented in the pipeline/alert.go comments. As any comments in that package become the auto generated docs for Kapacitor.